# Pandas

![](images/panda.jpg)

No, not *that* type of panda.

Pandas, derived from "panel data", is by far one of the most important tools on the belt of a Data Scientist. Pandas is a high level data-manipulation tool that is built on the NumPy package. With Pandas, you can explore a dataset on a CSV (Comma Separated Values) file and:
 - Clean the data by removing missing values and filtering rows and columns by certain criteria
 - Calculate statistics and answer questions about the data
 - Visualize the data using packages such as Matplotlib
 - Store the cleaned, transformed data back into a CSV file

## Installation

In order to use Pandas, you'll first need to install it. There are three different ways to install Pandas: PC and Mac users can run the cell below by clicking on it and selecting run (or Shift+Enter).

In [1]:
!pip install pandas

You are using pip version 19.0.3, however version 19.2.3 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


Alternatively, you can open up your terminal or command line and install it using either "conda install pandas" or "pip install pandas". If all goes according to plan, you should be able to import pandas below without any errors!

In [2]:
import pandas as pd

## DataFrames

Pandas' key data structure, the __DataFrame__, allows you to store and manipulate tabular data in rows of observations and columns of variables. A DataFrame is a multi-dimensional table made up of a collection of __Series__, or columns.

<img src="images/dataframe.png" alt="DataFrame" width="600"/>

### How to make a DataFrame

There are multiple ways to make a DataFrame. We'll start with the simplest option first, a dictionary, in order to illustrate the format of a DataFrame before moving on to different ways of creating DataFrames.

In [14]:
data = {
    'apples': [3, 2, 0, 1, 2],
    'oranges': [0, 3, 7, 2, 3]
}

purchases = pd.DataFrame(data)

purchases

apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2
4       2        3

*Voilà!* Creating a DataFrame is as simple as that. But how did we start with a dictionary and end up with a ```DataFrame```? After passing ```data``` into the DataFrame constructor, each key-value item within ```data``` corresponds to a column in the DataFrame. Let's take this a step further and change the indices to customer names:

In [15]:
purchases = pd.DataFrame(data, index=['Nick', 'Meekail', 'Quinn', 'Sonia', 'Jeff'])

purchases

apples  oranges
Nick          3        0
Meekail       2        3
Quinn         0        7
Sonia         1        2
Jeff          2        3

### Locating specific elements

Now, if we wanted to list out the purchases of a specific name, we would use ```loc```. This will tell us how many apples and oranges Meekail has as well as the datatype of the cells.

In [17]:
purchases.loc['Meekail']

apples     2
oranges    3
Name: Meekail, dtype: int64

Or, if you wanted to edit a specific cell using ```loc```, you'd specify the column after the row:

In [33]:
purchases.loc['Nick', 'oranges'] = 21

purchases

apples  oranges
Nick          3       21
Meekail       2        3
Quinn         0        7
Sonia         1        2
Jeff          2        3

Next, lets look at a slightly more complex example involving NumPy! Given two arrays, we'll pass them both into another array within the DataFrame constructor. Notice the ```T``` at the end of the DataFrame constructor; this means that we're transposing, or flipping the rows and columns so that 2x3 becomes 3x2.

In [49]:
import numpy as np

arr1 = np.array([2, 4, 6])
arr2 = np.array([3, 6, 9])

dataset = pd.DataFrame([arr1, arr2]).T

print(dataset.shape)
dataset

(3, 2)


0  1
0  2  3
1  4  6
2  6  9

If you want to modify the columns and rows after creating the DataFrame, you can call one of DataFrame's attributes ```columns``` and ```index``` respectively.

In [48]:
dataset.columns = ['apples', 'oranges']
dataset.index = ['Nick', 'Meekail', 'Quinn']

dataset

apples  oranges
Nick          2        3
Meekail       4        6
Quinn         6        9

Say you have three arrays of size n that you want to combine. Using hstack, you can stack the arrays and pass the resulting 2D array into the DataFrame constructor.

In [53]:
arr1, arr2, arr3 = np.zeros((3, 100, 1)) # Creates three 100x1 zero arrays

zero_df = pd.DataFrame(np.hstack((arr1, arr2, arr3)))

zero_df.head()

0    1    2
0  0.0  0.0  0.0
1  0.0  0.0  0.0
2  0.0  0.0  0.0
3  0.0  0.0  0.0
4  0.0  0.0  0.0

Additionally you can convert your DataFrame back into a NumPy array if need be.

In [57]:
zero_arr = zero_df.values

zero_arr

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0

## Experimenting with the IMDB Movie Dataset

Next, we'll move on to a more powerful option of loading datasets in Pandas. The ```read_csv()``` function will take in a CSV file and return a ```DataFrame```. If we take a look at the parameters of ```read_csv()```, we can see that the first parameter loads in the CSV file from the same folder our jupyter notebook file is in. The ```index_col``` parameter sets the index column header to "Title." The ```head()``` function can be used to return the first 5 rows of our dataset, or any number of rows by passing a number into the function. It is a good idea to view the first five or so rows to see what the dataset looks like.

In [58]:
movies_df = pd.read_csv("IMDB-Movie-Data.csv", index_col="Title")

movies_df.head(10)

Rank                       Genre  \
Title                                                       
Guardians of the Galaxy     1     Action,Adventure,Sci-Fi   
Prometheus                  2    Adventure,Mystery,Sci-Fi   
Split                       3             Horror,Thriller   
Sing                        4     Animation,Comedy,Family   
Suicide Squad               5    Action,Adventure,Fantasy   
The Great Wall              6    Action,Adventure,Fantasy   
La La Land                  7          Comedy,Drama,Music   
Mindhorn                    8                      Comedy   
The Lost City of Z          9  Action,Adventure,Biography   
Passengers                 10     Adventure,Drama,Romance   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
The Great Wall           European mercenaries searching for black powde...   
La La Land               A jazz pianist falls for an aspiring actress i...   
Mindhorn                 A has-been actor best known for playing the ti...   
The Lost City of Z       A true-life drama, centering on British explor...   
Passengers               A spacecraft traveling to a distant colony pla...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
The Great Wall                    Yimou Zhang   
La La Land                    Damien Chazelle   
Mindhorn                           Sean Foley   
The Lost City of Z                 James Gray   
Passengers                      Morten Tyldum   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
The Great Wall               Matt Damon, Tian Jing, Willem Dafoe, Andy Lau   
La La Land               Ryan Gosling, Emma Stone, Rosemarie DeWitt, J....   
Mindhorn                 Essie Davis, Andrea Riseborough, Julian Barrat...   
The Lost City of Z       Charlie Hunnam, Robert Pattinson, Sienna Mille...   
Passengers               Jennifer Lawrence, Chris Pratt, Michael Sheen,...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
The Great Wall           2016                103     6.1   56036   
La La Land               2016                128     8.3  258682   
Mindhorn                 2016                 89     6.4    2490   
The Lost City of Z       2016                141     7.1    7188   
Passengers               2016                116     7.0  192177   

            

Similarly, we could use the function ```tail()``` to list the last five rows.

In [59]:
movies_df.tail()

Rank                  Genre  \
Title                                                 
Secret in Their Eyes     996    Crime,Drama,Mystery   
Hostel: Part II          997                 Horror   
Step Up 2: The Streets   998    Drama,Music,Romance   
Search Party             999       Adventure,Comedy   
Nine Lives              1000  Comedy,Family,Fantasy   

                                                              Description  \
Title                                                                       
Secret in Their Eyes    A tight-knit team of rising investigators, alo...   
Hostel: Part II         Three American college students studying abroa...   
Step Up 2: The Streets  Romantic sparks occur between two dance studen...   
Search Party            A pair of friends embark on a mission to reuni...   
Nine Lives              A stuffy businessman finds himself trapped ins...   

                                Director  \
Title                                      
Secret in Their Eyes           Billy Ray   
Hostel: Part II                 Eli Roth   
Step Up 2: The Streets        Jon M. Chu   
Search Party              Scot Armstrong   
Nine Lives              Barry Sonnenfeld   

                                                                   Actors  \
Title                                                                       
Secret in Their Eyes    Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II         Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets  Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party            Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives              Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                        Year  Runtime (Minutes)  Rating  Votes  \
Title                                                            
Secret in Their Eyes    2015                111     6.2  27585   
Hostel: Part II         2007                 94     5.5  73152   
Step Up 2: The Streets  2008                 98     6.2  70699   
Search Party            2014                 93     5.6   4881   
Nine Lives              2016                 87     5.3  12435   

                        Revenue (Millions)  Metascore  
Title                                                  
Secret in Their Eyes                   NaN       45.0  
Hostel: Part II                      17.54       46.0  
Step Up 2: The Streets               58.01       50.0  
Search Party                           NaN       22.0  
Nine Lives                           19.64       11.0

Another useful function is ```sort_values()```, which will sort the dataset according to one selected column in ascending order by default.

In [31]:
movies_df.sort_values('Rating', ascending=False)

Rank                       Genre  \
Title                                                                    
The Dark Knight                         55          Action,Crime,Drama   
Inception                               81     Action,Adventure,Sci-Fi   
Dangal                                 118      Action,Biography,Drama   
Interstellar                            37      Adventure,Drama,Sci-Fi   
Kimi no na wa                           97     Animation,Drama,Fantasy   
The Intouchables                       250      Biography,Comedy,Drama   
Whiplash                               134                 Drama,Music   
The Prestige                            65        Drama,Mystery,Sci-Fi   
The Departed                           100        Crime,Drama,Thriller   
Taare Zameen Par                       992          Drama,Family,Music   
The Dark Knight Rises                  125             Action,Thriller   
The Lives of Others                    477              Drama,Thriller   
WALL·E                                 635  Animation,Adventure,Family   
3 Idiots                               431                Comedy,Drama   
Koe no katachi                         862     Animation,Drama,Romance   
Django Unchained                       145               Drama,Western   
La La Land                               7          Comedy,Drama,Music   
Jagten                                 456                       Drama   
Up                                     500  Animation,Adventure,Comedy   
Inglourious Basterds                    78         Adventure,Drama,War   
Toy Story 3                            689  Animation,Adventure,Comedy   
Paint It Black                         479                       Drama   
Bahubali: The Beginning                 27      Action,Adventure,Drama   
PK                                     766        Comedy,Drama,Romance   
The Wolf of Wall Street                 83      Biography,Comedy,Crime   
Inside Out                             242  Animation,Adventure,Comedy   
Gran Torino                            646                       Drama   
Incendies                              714           Drama,Mystery,War   
El secreto de sus ojos                 743       Drama,Mystery,Romance   
Hacksaw Ridge                           17     Biography,Drama,History   
...                                    ...                         ...   
Final Girl                             778             Action,Thriller   
Dead Awake                              28             Horror,Thriller   
Viking                                 523        Action,Drama,History   
Zoolander 2                            432                      Comedy   
The Boy Next Door                      556            Mystery,Thriller   
Wolves at the Door                     820             Horror,Thriller   
Max Steel                              788     Action,Adventure,Family   
Only for One Night                     747                    Thriller   
Shut In                                827       Drama,Horror,Thriller   
The Assignment                         772       Action,Crime,Thriller   
The Human Centipede (First Sequence)   937                      Horror   
Sex and the City 2                     953        Comedy,Drama,Romance   
Yoga Hosers                            801       Comedy,Fantasy,Horror   
Movie 43                               509              Comedy,Romance   
Fantastic Four                         553     Action,Adventure,Sci-Fi   
Exposed                                839         Crime,Drama,Mystery   
The Last Airbender                     582     Action,Adventure,Family   
Fifty Shades of Grey                    64      Drama,Romance,Thriller   
2307: Winter's Dream                   617                      Sci-Fi   
The Disappointments Room               238       Drama,Horror,Thriller   
The Black Room                         402                      Horror   
Birth of the Dragon                    526

### Finding the essential details in the data

```info()```  is an important function that provides essential details about the dataset, such as the column names, the shape of the dataset, the number of non-null values, the type of data in each column, and the DataFrame's memory usage. ```info()``` should be one of the first commands you run after loading a dataset since it will give you an idea of the dataset in a brief summary.

In [60]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
Rank                  1000 non-null int64
Genre                 1000 non-null object
Description           1000 non-null object
Director              1000 non-null object
Actors                1000 non-null object
Year                  1000 non-null int64
Runtime (Minutes)     1000 non-null int64
Rating                1000 non-null float64
Votes                 1000 non-null int64
Revenue (Millions)    872 non-null float64
Metascore             936 non-null float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


### Handling duplicates

Datasets you interact with won't always be as clean as this one. Often you will find datasets with duplicates, NaNs, syntax errors, or just plain irrelevant data. This matters in the end once you've trained your machine learning algorithm on dirty data because you can typically expect less than optimal results on your test set. *Garbage in, garbage out*, as they say. We'll first take a look at removing duplicates by using ```append()``` to add another movie DataFrame to our DataFrame already.

In [61]:
temp_df = movies_df.append(movies_df)

temp_df.shape

(2000, 11)

It should be clear that the number of rows has now doubled after appending the same dataset to it. Removing these duplicates is made extremely simple by Pandas because you can just call ```drop_duplicates()``` and Pandas will do the work for you! Notice how the shape of the dataset returns to 1000 rows by 11 columns.

In [62]:
temp_df = temp_df.drop_duplicates()

temp_df.shape

(1000, 11)

From looking at this info, notice that the dataset has missing values in the ```Revenue``` and ```Metascore``` columns. Say that you don't want to keep reassigning a DataFrame to the same variable. The ```inplace``` argument will take care of that. The ```inplace``` parameter also applies to other DataFrame functions, as you will see. ```keep``` is another parameter that can be used to choose which duplicate to keep and which duplicates to drop. There are three arguments for ```keep```:

 - ```first```: (default) Drop duplicates except first occurrence.
 - ```last```: Drop duplicates except last occurrence.
 - ```False```: Drop all duplicates.

In [64]:
temp_df = movies_df.append(movies_df)

temp_df.drop_duplicates(inplace=True, keep=False)

temp_df.shape

(0, 11)

### Cleaning up columns

Let's take a look at our columns one more time using the ```columns``` attribute:

In [65]:
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

Everything looks good, except for two columns: ```Runtime (Minutes)``` and ```Revenue (Millions)```. *Why are these two columns problematic?*

In [66]:
movies_df["Runtime(Minutes)"]

KeyError: 'Runtime(Minutes)'

In [67]:
movies_df["Rank"]

Title
Guardians of the Galaxy                       1
Prometheus                                    2
Split                                         3
Sing                                          4
Suicide Squad                                 5
The Great Wall                                6
La La Land                                    7
Mindhorn                                      8
The Lost City of Z                            9
Passengers                                   10
Fantastic Beasts and Where to Find Them      11
Hidden Figures                               12
Rogue One                                    13
Moana                                        14
Colossal                                     15
The Secret Life of Pets                      16
Hacksaw Ridge                                17
Jason Bourne                                 18
Lion                                         19
Arrival                                      20
Gold                              

These columns are problematic because referencing them with parenthesis in their names will throw a ```Key Error```. The solution to this is to rename them using the ```rename()``` function. We can specify which columns we want to change by passing a dictionary into the ```columns``` parameter with the current column names as  the key and the new column names as the value.

In [68]:
movies_df.rename(columns={
    'Runtime (Minutes)': 'Runtime',
    'Revenue (Millions)': 'Revenue_millions'
}, inplace=True)

movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue_millions', 'Metascore'],
      dtype='object')

Cleaning up the columns can be taken a step further by converting each column name to lower case! The columns array can be set to a new array containing the same exact columns, except we iterate through them first and apply ```col.lower()``` to convert the names to lowercase.

It's always a good idea to lowercase, remove special characters, and replace spaces with underscores any column names in a dataset if you intend to work with a dataset for a while.

In [69]:
movies_df.columns = [col.lower() for col in movies_df]

movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

### Working with missing values

Missing or null values are common in most datasets that you'll encounter. There are two options in dealing with nulls:

1. Get rid of rows or columns with nulls
2. Replace nulls with non-null values, a technique known as __imputation__. 

First, let's check how many nulls are in each column of our dataset:

In [70]:
movies_df.isnull()

rank  genre  description  director  \
Title                                                                          
Guardians of the Galaxy                  False  False        False     False   
Prometheus                               False  False        False     False   
Split                                    False  False        False     False   
Sing                                     False  False        False     False   
Suicide Squad                            False  False        False     False   
The Great Wall                           False  False        False     False   
La La Land                               False  False        False     False   
Mindhorn                                 False  False        False     False   
The Lost City of Z                       False  False        False     False   
Passengers                               False  False        False     False   
Fantastic Beasts and Where to Find Them  False  False        False     False   
Hidden Figures                           False  False        False     False   
Rogue One                                False  False        False     False   
Moana                                    False  False        False     False   
Colossal                                 False  False        False     False   
The Secret Life of Pets                  False  False        False     False   
Hacksaw Ridge                            False  False        False     False   
Jason Bourne                             False  False        False     False   
Lion                                     False  False        False     False   
Arrival                                  False  False        False     False   
Gold                                     False  False        False     False   
Manchester by the Sea                    False  False        False     False   
Hounds of Love                           False  False        False     False   
Trolls                                   False  False        False     False   
Independence Day: Resurgence             False  False        False     False   
Paris pieds nus                          False  False        False     False   
Bahubali: The Beginning                  False  False        False     False   
Dead Awake                               False  False        False     False   
Bad Moms                                 False  False        False     False   
Assassin's Creed                         False  False        False     False   
...                                        ...    ...          ...       ...   
Texas Chainsaw 3D                        False  False        False     False   
Disturbia                                False  False        False     False   
Rock of Ages                             False  False        False     False   
Scream 4                                 False  False        False     False   
Queen of Katwe                           False  False        False     False   
My Big Fat Greek Wedding 2               False  False        False     False   
Dark Places                              False  False        False     False   
Amateur Night                            False  False        False     False   
It's Only the End of the World           False  False        False     False   
The Skin I Live In                       False  False        False     False   
Miracles from Heaven                     False  False        False     False   
Annie                                    False  False        False     False   
Across the Universe                      False  False        False     False   
Let's Be Cops                            False  False        False     False   
Max                                      False  False        False     False   
Your Highness                            False  False        False     False   
Final Destination 5                      False  False        False     False   
Endless Love                             F

```isnull()``` is good for showing which cells have null values (listed as True), but it's not helpful in giving us an idea of *how many* cells are null. What we can do is count the number of nulls in each column:

In [71]:
movies_df.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

From this, we can now see that our problem area lies within ```revenue_millions``` and ```metascore```, which has 128 missing values and 64 missing values respectively. There are two options now: __removing__ the null values, or __filling__ the null values. First, let's look at how to remove rows with null values.

In [75]:
movies_df.dropna()

rank                       genre  \
Title                                                                       
Guardians of the Galaxy                     1     Action,Adventure,Sci-Fi   
Prometheus                                  2    Adventure,Mystery,Sci-Fi   
Split                                       3             Horror,Thriller   
Sing                                        4     Animation,Comedy,Family   
Suicide Squad                               5    Action,Adventure,Fantasy   
The Great Wall                              6    Action,Adventure,Fantasy   
La La Land                                  7          Comedy,Drama,Music   
The Lost City of Z                          9  Action,Adventure,Biography   
Passengers                                 10     Adventure,Drama,Romance   
Fantastic Beasts and Where to Find Them    11    Adventure,Family,Fantasy   
Hidden Figures                             12     Biography,Drama,History   
Rogue One                                  13     Action,Adventure,Sci-Fi   
Moana                                      14  Animation,Adventure,Comedy   
Colossal                                   15         Action,Comedy,Drama   
The Secret Life of Pets                    16  Animation,Adventure,Comedy   
Hacksaw Ridge                              17     Biography,Drama,History   
Jason Bourne                               18             Action,Thriller   
Lion                                       19             Biography,Drama   
Arrival                                    20        Drama,Mystery,Sci-Fi   
Gold                                       21    Adventure,Drama,Thriller   
Manchester by the Sea                      22                       Drama   
Trolls                                     24  Animation,Adventure,Comedy   
Independence Day: Resurgence               25     Action,Adventure,Sci-Fi   
Bad Moms                                   29                      Comedy   
Assassin's Creed                           30      Action,Adventure,Drama   
Why Him?                                   31                      Comedy   
Nocturnal Animals                          32              Drama,Thriller   
X-Men: Apocalypse                          33     Action,Adventure,Sci-Fi   
Deadpool                                   34     Action,Adventure,Comedy   
Resident Evil: The Final Chapter           35        Action,Horror,Sci-Fi   
...                                       ...                         ...   
That Awkward Moment                       956              Comedy,Romance   
Legion                                    957       Action,Fantasy,Horror   
End of Watch                              958        Crime,Drama,Thriller   
3 Days to Kill                            959       Action,Drama,Thriller   
Lucky Number Slevin                       960         Crime,Drama,Mystery   
Trance                                    961         Crime,Drama,Mystery   
Into the Forest                           962       Drama,Sci-Fi,Thriller   
The Other Boleyn Girl                     963     Biography,Drama,History   
I Spit on Your Grave                      964       Crime,Horror,Thriller   
Texas Chainsaw 3D                         971             Horror,Thriller   
Rock of Ages                              973        Comedy,Drama,Musical   
Scream 4                                  974              Horror,Mystery   
Queen of Katwe                            975       Biography,Drama,Sport   
My Big Fat Greek Wedding 2                976       Comedy,Family,Romance   
The Skin I Live In                        980              Drama,Thriller   
Miracles from Heaven                      981      Biography,Drama,Family   
Annie                                     982         Comedy,Drama,Family   
Across the Universe                       983       Drama,Fantasy,Musical   
Let's Be Cops                             984                      Comedy   
Max                                  

By calling ```dropna()```, any row with at least one null value will be deleted. Notice how the above code doesn't change our ```movies_df``` variable since ```inplace``` was not specified. At the bottom of the dataframe, it shows that there are 838 rows left. 

If instead you wanted to drop a column consisting of null values instead of rows, you could specify ```axis=1``` in ```dropna()```.

In [76]:
movies_df.dropna(axis=1)

rank                       genre  \
Title                                                                       
Guardians of the Galaxy                     1     Action,Adventure,Sci-Fi   
Prometheus                                  2    Adventure,Mystery,Sci-Fi   
Split                                       3             Horror,Thriller   
Sing                                        4     Animation,Comedy,Family   
Suicide Squad                               5    Action,Adventure,Fantasy   
The Great Wall                              6    Action,Adventure,Fantasy   
La La Land                                  7          Comedy,Drama,Music   
Mindhorn                                    8                      Comedy   
The Lost City of Z                          9  Action,Adventure,Biography   
Passengers                                 10     Adventure,Drama,Romance   
Fantastic Beasts and Where to Find Them    11    Adventure,Family,Fantasy   
Hidden Figures                             12     Biography,Drama,History   
Rogue One                                  13     Action,Adventure,Sci-Fi   
Moana                                      14  Animation,Adventure,Comedy   
Colossal                                   15         Action,Comedy,Drama   
The Secret Life of Pets                    16  Animation,Adventure,Comedy   
Hacksaw Ridge                              17     Biography,Drama,History   
Jason Bourne                               18             Action,Thriller   
Lion                                       19             Biography,Drama   
Arrival                                    20        Drama,Mystery,Sci-Fi   
Gold                                       21    Adventure,Drama,Thriller   
Manchester by the Sea                      22                       Drama   
Hounds of Love                             23          Crime,Drama,Horror   
Trolls                                     24  Animation,Adventure,Comedy   
Independence Day: Resurgence               25     Action,Adventure,Sci-Fi   
Paris pieds nus                            26                      Comedy   
Bahubali: The Beginning                    27      Action,Adventure,Drama   
Dead Awake                                 28             Horror,Thriller   
Bad Moms                                   29                      Comedy   
Assassin's Creed                           30      Action,Adventure,Drama   
...                                       ...                         ...   
Texas Chainsaw 3D                         971             Horror,Thriller   
Disturbia                                 972      Drama,Mystery,Thriller   
Rock of Ages                              973        Comedy,Drama,Musical   
Scream 4                                  974              Horror,Mystery   
Queen of Katwe                            975       Biography,Drama,Sport   
My Big Fat Greek Wedding 2                976       Comedy,Family,Romance   
Dark Places                               977      Drama,Mystery,Thriller   
Amateur Night                             978                      Comedy   
It's Only the End of the World            979                       Drama   
The Skin I Live In                        980              Drama,Thriller   
Miracles from Heaven                      981      Biography,Drama,Family   
Annie                                     982         Comedy,Drama,Family   
Across the Universe                       983       Drama,Fantasy,Musical   
Let's Be Cops                             984                      Comedy   
Max                                       985            Adventure,Family   
Your Highness                             986    Adventure,Comedy,Fantasy   
Final Destination 5                       987             Horror,Thriller   
Endless Love                              988               Drama,Romance   
Martyrs                                   989                      Horror   
Selma                                

At the bottom, it now shows that there are 9 columns instead of the original 11. ```axis=1``` may not make sense in how it switches from rows being deleted to columns being deleted, but the idea behind it lies within the shape of the dataframe. The shape of the dataframe is a tuple, in which each axis represents an index of the tuple. So, index 0 equals the rows (1000) and index 1 equals the columns (11).

In [77]:
movies_df.shape

(1000, 11)

While dropping null values can prove useful in specific cases, it would be a waste to drop these rows entirely since it's perfectly good data that could be used in training a machine learning algorithm. This is where __imputation__ comes in handy. Imputation is a feature engineering technique used to replace missing data with substituted values, generally either the __mean__ or __median__ of that column. Let's impute the missing values in the ```revenue_millions``` column by creating a variable for it and finding the mean of it.

In [82]:
revenue = movies_df['revenue_millions']

revenue_mean = revenue.mean()

revenue_mean

82.95637614678897

Now that we have the mean, we can fill null values in the revenue_millions column using ```fillna()```:

In [84]:
revenue.fillna(revenue_mean, inplace=True)

movies_df.isnull().sum()

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

As you can see, we've successfully filled all null values for the ```revenue_millions``` column! Notice that we changed the original ```movies_df``` by using ```inplace=True```. Imputing the entire column with the same mean value is a simple way to fill null values.

__Question:__ Is there a better, more specific imputation we could apply to the revenue column by using mean?

### Understanding variables

Now let's get an idea of what functions you can use to deepen your understanding of the dataset. First, we'll look at ```describe()``` to get a summary of the distribution of continue variables in the dataset. ```describe()``` can be useful in figuring out what type of plot to use to visualize your data.

In [85]:
movies_df.describe()

rank         year      runtime       rating         votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       revenue_millions   metascore  
count       1000.000000  936.000000  
mean          82.956376   58.985043  
std           96.412043   17.194757  
min            0.000000   11.000000  
25%           17.442500   47.000000  
50%           60.375000   59.500000  
75%           99.177500   72.000000  
max          936.630000  100.000000

```describe()``` can also be applied to a categorical variable to get a count of rows, unique count of categories, top category, and the frequency of the top category:

In [87]:
movies_df['genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

We can see from the above description that of the 207 unique categories, Action/Adventure/Sci-Fi is the most popular category with 50 occurrences.

If you want to get a frequency of the different categories, you can use ```value_counts()``` to list out the frequency of all values in a column:

In [88]:
movies_df['genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Action,Adventure,Fantasy      27
Animation,Adventure,Comedy    27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

### DataFrame functions recap

We just flew through multiple functions used to provide insight into our dataset and modify properties of the dataset. Here's a recap of all these functions:

 - ```head()``` returns the first 5 elements of the dataset, unless a specific number is passed as a parameter.
 - ```tail()``` returns the last 5 elements of the dataset, unless a specific number is passed as a parameter.
 - ```sort_values()``` returns the dataset sorted according to a specific column in ascending order by default.
 - ```info()``` returns details about the dataset, such as the column names, the shape of the dataset, the number of non-null values, the type of data in each column, and the DataFrame's memory usage.
 - ```append()``` returns a DataFrame with a new DataFrame appended to the previous DataFrame.
 - ```drop_duplicates()``` returns a DataFrame with any subsequent duplicates dropped from the dataset.
 - ```rename()``` renames axes labels.
 - ```isnull()``` returns a DataFrame showing which cells have null values (listed as True).
 - ```dropna()``` returns a DataFrame after deleting rows (or columns) with at least one null value.
 - ```fillna()``` returns a DataFrame with all null values replaced with a non-null value.
 - ```describe()``` returns a DataFrame with descriptive statistics that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding NaN values.
 
 For more info on data cleaning, come out to our Introduction to Data Analytics workshop, or check out this link for a quick overview: https://elitedatascience.com/data-cleaning

### DataFrame manipulation

Finally, we'll cover methods for slicing, selecting, and extracting. Let's take a look at extracting first.

### Extracting

In [89]:
genre_col = movies_df['genre']

type(genre_col)

pandas.core.series.Series

Unsurprisingly, the single column we selected from movies_df is a *Series* object. 

But what if you want a single-column DataFrame? To do this, you need to pass a *list* of column names:

In [91]:
genre_col = movies_df[['genre']]

type(genre_col)

pandas.core.frame.DataFrame

Creating a subset of columns is as easy as adding to the list of columns we want from ```movies_df```:

In [93]:
subset = movies_df[['genre', 'rating']]

subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

Now for retrieving rows, there are two options:
 - ```.loc```: __loc__ates by name
 - ```.iloc```: __loc__ates by numerical index

In [94]:
prom = movies_df.loc["Prometheus"]

prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                              7
votes                                                          485820
revenue_millions                                               126.46
metascore                                                          65
Name: Prometheus, dtype: object

Alternatively, we can select Prometheus using its numerical index:

In [95]:
prom = movies_df.iloc[1]

### Slicing

```loc``` and ```iloc``` can be thought as similar to python ```list``` slicing. We can show this by selecting multiple rows:

In [96]:
# Lines 3 and 5 are equivalent

movie_subset = movies_df.loc['Prometheus': 'Sing'] # Select rows from Prometheus to Sing

movie_subset = movies_df.iloc[1:4] # Select indices 1 through 3

movie_subset

rank                     genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       actors  year  runtime  \
Title                                                                          
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012      124   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016      117   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016      108   

            rating   votes  revenue_millions  metascore  
Title                                                    
Prometheus     7.0  485820            126.46       65.0  
Split          7.3  157606            138.12       62.0  
Sing           7.2   60545            270.32       59.0

Slicing with ```.iloc``` is the same as slicing with lists, the object at the index at the end is not included.

### Conditional Selections

Now that we've gone over how to select columns and rows, what if we want to make a conditional selection?

For example, wjat of we want to filter our movies DataFrame to show only films directed by Ridley Scott or films with a rating greater than or equal to 8.0?

To do that, we take a column from the DataFrame and apply a Boolean condition to it.

In [97]:
movies_df[movies_df['director'] == "Ridley Scott"]

rank                     genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   
The Counselor            522      Crime,Drama,Thriller   
A Good Year              531      Comedy,Drama,Romance   
Body of Lies             738      Action,Drama,Romance   

                                                              description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
The Martian             An astronaut becomes stranded on Mars after hi...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   
The Counselor           A lawyer finds himself in over his head when h...   
A Good Year             A British investment broker inherits his uncle...   
Body of Lies            A CIA agent on the ground in Jordan hunts down...   

                            director  \
Title                                  
Prometheus              Ridley Scott   
The Martian             Ridley Scott   
Robin Hood              Ridley Scott   
American Gangster       Ridley Scott   
Exodus: Gods and Kings  Ridley Scott   
The Counselor           Ridley Scott   
A Good Year             Ridley Scott   
Body of Lies            Ridley Scott   

                                                                   actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   
The Counselor           Michael Fassbender, Penélope Cruz, Cameron Dia...   
A Good Year             Russell Crowe, Abbie Cornish, Albert Finney, M...   
Body of Lies            Leonardo DiCaprio, Russell Crowe, Mark Strong,...   

                        year  runtime  rating   votes  revenue_millions  \
Title                                                                     
Prometheus              2012      124     7.0  485820            126.46   
The Martian             2015      144     8.0  556097            228.43   
Robin Hood              2010      140     6.7  221117            105.22   
American Gangster       2007      157     7.8  337835            130.13   
Exodus: Gods and Kings  2014      150     6.0  137299             65.01   
The Counselor           2013      117     5.3   84927             16.97   
A Good Year             2006      117     6.9   74674              7.46   
Body of Lies            2008      128     7.1  182305             39.38   

                        metascore  
Title                              
Prometheus                   65.0  
The Martian                  80.0  
Robin Hood                   53.0  
American Gangster            76.0  
Exodus: Gods and Kings       52.0  
The Counselor                48.0  
A Good Year                  47.0  
Body of Lies                 57.0

If we want to filter the DataFrame by ratings greater than 8.6, we could do:

In [98]:
movies_df[movies_df['rating'] >= 8.6].head(3)

rank                    genre  \
Title                                            
Interstellar       37   Adventure,Drama,Sci-Fi   
The Dark Knight    55       Action,Crime,Drama   
Inception          81  Action,Adventure,Sci-Fi   

                                                       description  \
Title                                                                
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
Inception        A thief, who steals corporate secrets through ...   

                          director  \
Title                                
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
Inception        Christopher Nolan   

                                                            actors  year  \
Title                                                                      
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 runtime  rating    votes  revenue_millions  metascore  
Title                                                                   
Interstellar         169     8.6  1047747            187.99       74.0  
The Dark Knight      152     9.0  1791916            533.32       82.0  
Inception            148     8.8  1583625            292.57       74.0

We can narrow our selections by using the logical operators ```|``` for "or" and ```&``` for "and".

If we want to filter the DataFrame to show only movies by Christopher Nolan OR Ridley Scott we would write:

In [99]:
movies_df[(movies_df['director'] == 'Christopher Nolan') | (movies_df['director'] == 'Ridley Scott')].head()

rank                     genre  \
Title                                             
Prometheus          2  Adventure,Mystery,Sci-Fi   
Interstellar       37    Adventure,Drama,Sci-Fi   
The Dark Knight    55        Action,Crime,Drama   
The Prestige       65      Drama,Mystery,Sci-Fi   
Inception          81   Action,Adventure,Sci-Fi   

                                                       description  \
Title                                                                
Prometheus       Following clues to the origin of mankind, a te...   
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
The Prestige     Two stage magicians engage in competitive one-...   
Inception        A thief, who steals corporate secrets through ...   

                          director  \
Title                                
Prometheus            Ridley Scott   
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
The Prestige     Christopher Nolan   
Inception        Christopher Nolan   

                                                            actors  year  \
Title                                                                      
Prometheus       Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
The Prestige     Christian Bale, Hugh Jackman, Scarlett Johanss...  2006   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 runtime  rating    votes  revenue_millions  metascore  
Title                                                                   
Prometheus           124     7.0   485820            126.46       65.0  
Interstellar         169     8.6  1047747            187.99       74.0  
The Dark Knight      152     9.0  1791916            533.32       82.0  
The Prestige         130     8.5   913152             53.08       66.0  
Inception            148     8.8  1583625            292.57       74.0

How about if we want all movies released between 2005 and 2010, have a rating above 8.0, but made below the 25th percentile in revenue?

In [100]:
movies_df[
    ((movies_df['year'] >= 2005) & (movies_df['year'] <= 2010))
    & (movies_df['rating'] > 8.0)
    & (movies_df['revenue_millions'] < movies_df['revenue_millions'].quantile(0.25))
]

rank               genre  \
Title                                           
3 Idiots              431        Comedy,Drama   
The Lives of Others   477      Drama,Thriller   
Incendies             714   Drama,Mystery,War   
Taare Zameen Par      992  Drama,Family,Music   

                                                           description  \
Title                                                                    
3 Idiots             Two friends are searching for their long lost ...   
The Lives of Others  In 1984 East Berlin, an agent of the secret po...   
Incendies            Twins journey to the Middle East to discover t...   
Taare Zameen Par     An eight-year-old boy is thought to be a lazy ...   

                                             director  \
Title                                                   
3 Idiots                              Rajkumar Hirani   
The Lives of Others  Florian Henckel von Donnersmarck   
Incendies                            Denis Villeneuve   
Taare Zameen Par                           Aamir Khan   

                                                                actors  year  \
Title                                                                          
3 Idiots               Aamir Khan, Madhavan, Mona Singh, Sharman Joshi  2009   
The Lives of Others  Ulrich Mühe, Martina Gedeck,Sebastian Koch, Ul...  2006   
Incendies            Lubna Azabal, Mélissa Désormeaux-Poulin, Maxim...  2010   
Taare Zameen Par     Darsheel Safary, Aamir Khan, Tanay Chheda, Sac...  2007   

                     runtime  rating   votes  revenue_millions  metascore  
Title                                                                      
3 Idiots                 170     8.4  238789              6.52       67.0  
The Lives of Others      137     8.5  278103             11.28       89.0  
Incendies                131     8.2   92863              6.86       80.0  
Taare Zameen Par         165     8.5  102697              1.20       42.0

Source Material: https://www.learndatasci.com/tutorials/python-pandas-tutorial-complete-introduction-for-beginners/